File: Extensions/Tessa.Extensions.Client/Files/DocxToPdfFileExtension.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Tessa.Cards;
using Tessa.FileConverters;
using Tessa.Files;
using Tessa.Platform.Collections;
using Tessa.Platform.Storage;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Files;
using Tessa.UI.Menu;

using Unity;
using Unity.Lifetime;

namespace Tessa.Extensions.Client.Files
{
    // нужно еще расширение файла docx отлавливать и открывать только на нем
    public sealed class DocxToPdfFileExtension : FileExtension
    {
        private readonly ICardRepository cardRepository;
        public DocxToPdfFileExtension(ICardRepository cardRepository)
        {
            this.cardRepository = cardRepository;
        }

        public override async Task OpeningMenu(IFileExtensionContext context)
        {
            ICardEditorModel editor = UIContext.Current.CardEditor;
            var uiContext = UIContext.Current.CardEditor.Context;
            ICardModel model;
            if (editor == null
                || (model = editor.CardModel) == null
                //|| model.CardType.Name != ""
                )
            {
                return;
            }

            // Добавляем пункт меню для получения файлов
            context.Actions.AddRange(
                new MenuAction(
                    "DocxToPdf",
                    "Конвертировать в pdf",
                    context.Icons.Get("Thin427"),
                    new DelegateCommand(async o => ConvertDocxFileToPdfMultiple(cardRepository, editor, context.Control, context.Files))
                    //,
                    //isSelectable: true,
                    //isSelected: isContainsExternalFiles
                    )
            );
        }

        // convert
        private static async Task ConvertDocxFileToPdfMultiple(ICardRepository cardRepository, ICardEditorModel editor, IFileControl control, IFileCollection files)
        {
            using (var splash = TessaSplash.Create("Конвертация файлов в pdf..."))
            {
                Guid cardID = editor.CardModel.Card.ID;
                string notConvertedDocxFiles = "";
                string notSavedDocxFiles = "";

                for (int i = 0; i < files.Count; i++)
                {
                    int versionsCount = files[i].Versions.Count;
                    if (files[i].Versions[versionsCount - 1].State == Tessa.Files.FileVersionState.Created)
                    {
                        if (!string.IsNullOrEmpty(notSavedDocxFiles))
                            notSavedDocxFiles += ", ";
                        else notSavedDocxFiles += (files[i].Name + " ");
                    }
                }
                if (!string.IsNullOrEmpty(notSavedDocxFiles))
                {
                    TessaDialog.ShowMessage($"Файлы: {notSavedDocxFiles} добавлены в карточку, но не сохранены. Сохраните карточку перед конвертацией файлов.", "Конвертация в pdf");
                    return;
                }

                for (int i = 0; i < files.Count; i++)
                {
                    int versionsCount = files[i].Versions.Count;
                    Guid versionRowID = files[i].Versions[versionsCount - 1].ID;

                    CardRequest request = new CardRequest
                    {
                        RequestType = Shared.PnrRequestTypes.GetPdfFromDocxTypeID,
                        Info =
                    {
                        { "versionRowID", versionRowID },
                        { "cardID", cardID },
                        { "fileName", files[i].Name }
                    }
                    };

                    CardResponse response = await cardRepository.RequestAsync(request);

                    Tessa.Platform.Validation.ValidationResult result = response.ValidationResult.Build();
                    if (!result.IsSuccessful || !response.Info.Get<bool>("ConvertFileResult"))
                    {
                        if (!string.IsNullOrEmpty(notConvertedDocxFiles))
                            notConvertedDocxFiles += ", ";
                        else notConvertedDocxFiles += (files[i].Name + " ");
                    }
                }

                if (!string.IsNullOrEmpty(notConvertedDocxFiles))
                {
                    TessaDialog.ShowError($"Конвертация файлов: {notConvertedDocxFiles}в pdf завершилась с ошибкой.", "Конвертация в pdf");
                }

                await editor.RefreshCardAsync(editor.Context);
            }
        }
    }
}

File: WebClient SDK/src/PnrScripts/Files/DocxToPdfFileExtension.ts
import { FileExtension, IFileExtensionContext } from 'tessa/ui/files';
import { Card } from 'tessa/cards';
import { UIContext, MenuAction, showMessage, showLoadingOverlay } from 'tessa/ui';
import { ICardEditorModel } from 'tessa/ui/cards';
import { CardRequest, CardService } from 'tessa/cards/service';
import { FileViewModel } from 'tessa/ui/cards/controls';
import { DotNetType, createTypedField } from 'tessa/platform';
import { PnrRequestTypes } from '../Shared/PnrRequestTypes';

export class DocxToPdfFileExtension extends FileExtension {
  public async openingMenu(context: IFileExtensionContext) {
    let uiContext = UIContext.current;
    let editor = uiContext.cardEditor;
    let card: Card = context.control.model.card;
    // Добавляем пункт меню для получения файлов
    context.actions.push(
        new MenuAction(
          'DocxToPdf',
          'Конвертировать в pdf',
          "ta icon-thin-427",
          async () => await this.ConvertDocxFileToPdfMultiple(card, context.files, editor),
          null,
          false
      ));
  }

  private async ConvertDocxFileToPdfMultiple(card: Card, files: ReadonlyArray<FileViewModel>, editor: ICardEditorModel | null) {
    if (!editor) return;
    await showLoadingOverlay(async () => {
        let cardID = card.id;
        let notConvertedDocxFiles = '';
        let notSavedDocxFiles = '';
        for (let i = 0; i < files.length; i++) {
            let versionsCount = files[i].model.versions.length;
            let versionState = files[i].model.versions[versionsCount - 1].state;
            if (versionState == 0) {
                if (notSavedDocxFiles != null && notSavedDocxFiles.length > 0) {
                    notSavedDocxFiles += ', ';
                }
                else {
                    notSavedDocxFiles += (files[i].model.name + ' ');
                }
            }
        }
        if (notSavedDocxFiles != null && notSavedDocxFiles.length > 0) {
            showMessage('Файлы: ' + notSavedDocxFiles + ' добавлены в карточку, но не сохранены. Сохраните карточку перед конвертацией файлов.', 'Конвертация в pdf');
            return;
        }
    
        for (let i = 0; i < files.length; i++) {
            let versionsCount = files[i].model.versions.length;
            let versionRowID = files[i].model.versions[versionsCount - 1].id;
    
            let request = new CardRequest();
                request.requestType = PnrRequestTypes.GetPdfFromDocxTypeID;
                request.info["versionRowID"] = createTypedField(versionRowID, DotNetType.Guid);
                request.info["cardID"] = createTypedField(cardID, DotNetType.Guid);
                request.info["fileName"] = createTypedField(files[i].model.name, DotNetType.String);
    
            // проверка ответа
            let response = await CardService.instance.request(request);
            if (!response.validationResult.isSuccessful || !response.info.ConvertFileResult) {
                if (notConvertedDocxFiles != null && notConvertedDocxFiles.length > 0) {
                    notConvertedDocxFiles += ', ';
                }
                else {
                    notConvertedDocxFiles += (files[i].model.name + ' ');
                }
            }
        }
    
        if (notConvertedDocxFiles != null && notConvertedDocxFiles.length > 0) {
            showMessage('Конвертация файлов: ' + notConvertedDocxFiles + ' в pdf завершилась с ошибкой.', 'Конвертация в pdf');
            return;
        }
    });
    await editor.refreshCard(editor.context);
  }
}